src/actions/facebookcustomermatch/lib/executor.ts

```diff
@@ -0,0 +1,478 @@
+            // dimensions: Array(8), table_calculations: Array(0), pivots: Array(0)}
+            let combinedFields = [...fieldData.dimensions,
+                                  ...fieldData.measures,
+                                  ...fieldData.table_calculations]
+            // prune the object to just the subset of data we need, then combine into one object
+            combinedFields = combinedFields.reduce((aggregator, field) => {
+              aggregator[field.name] = {
+                value: null, // not important for determining schema
+                tags: field.tags,
+              }
+              return aggregator
+            }, {})
+            this.determineSchema(combinedFields)
+          }
+
+          return oboe.drop
+        }, "!.data.*" : (row: any) => {
+          // the reduce below just strips the superfluous object and "value property from the row
+          // i.e. { users.city:{value: 'Abbeville'}, "users.zip": ... }
+          // becomes
+          // { users.city: 'Abbeville' }
+          row = Object.entries(row).reduce((accumulator: any, [key, val]: any[]) => {
+            accumulator[key] = val.value
+            return accumulator
+          }, {})
+          this.handleRow(row)
+          this.scheduleBatch()
+          return oboe.drop
+        },
+        })
+        .done(() => {
+          this.scheduleBatch(true)
+          resolve()
+        })
+        .fail(reject)
+    })
+  }
+
+  private determineSchema(row: any) {
+    for (const columnLabel of Object.keys(row)) {
+      let tagMatched = false
+      for (const mapping of this.fieldMapping) {
+        const {fallbackRegex, lookMLFieldName} = mapping
+
+        // attempt to match fields by lookml tags first
+        if (row[columnLabel].tags && row[columnLabel].tags.length > 0) {
+          if (row[columnLabel].tags.some((tag: string) =>
+          tag.toLowerCase() === lookMLFieldName.toLowerCase())) {
+            tagMatched = true
+            this.schema[columnLabel] = mapping;
+            winston.debug("info", `Matched ${columnLabel} by LookML field tag.`)
+            break
+          }
+        }
+
+        if (columnLabel.match(fallbackRegex)) {
+          tagMatched = true
+          this.schema[columnLabel] = mapping
+          winston.debug("info", `Matched ${columnLabel} by regex.`)
+          break
+        }
+      }
+      if (!tagMatched) {
+        winston.debug("info", `Could not match field ${columnLabel}
+        by tags or regex. Dropping it from upload.`)
+      }
+    }
+    const formattedRow = this.getFormattedRow(row, this.schema)
+    this.matchedHashCombinations = this.getMatchingHashCombinations(formattedRow,
+      validFacebookHashCombinations)
+    if (this.matchedHashCombinations.length <= 0) {
+      throw new Error("Could not match your data to any valid Facebook data combinations." +
+      "See documentation for details. No data has been sent to Facebook. CLEANFAIL")
```

Comment:
I tested this with some data with no columns matched. Throwing the error here makes oboe fail and causes the AH server to crash. Is there another way we can handle this?

Pasting error for reference:
```
.../node_modules/oboe/dist/oboe-node.js:2041
          throw new Error(e.message)
    at Timeout._onTimeout (.../node_modules/oboe/dist/oboe-node.js:2041:17)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
error Command failed with exit code 1.
```

Reply:
Hmmm. The control flow is weird. We would need to kill the stream somehow then indicate to the top level `execute` that no columns were matched so it can send this 200 error.

Maybe we could pass a callback that aborts streaming because of no columns matching. Would need to see what the oboe api has to do that

Reply:
that or try/catch inside the stream function, kill the stream, THEN throw the error. so goofy

Reply:
okay, took another look. if you catch the error in the first function (just wrap `determineSchema(...)` in a try/catch) you can call `reject` with an error message/ or "clean fail" flag then send the response exactly like in `executor.js:203`

heck, rejecting the promise might even throw an error and then nothing needs to change